### ISSUE TITLE
Sensitive content report moderation workflow

### ISSUE THREAD

**Issue by obulat** [WordPress/openverse#383]
| Start Date | Project Lead    | Actual Ship Date |
| ---------- | ----------      | ---------------  | 
| 2023-02-13 | @sarayourfriend | TBD              |

## Description

This project creates the sensitive content reporting flow guidelines.

## Documents

- [x] #1176
- [x] #1965
- [x] #1966
- [x] #1967
- [x] ~#1968~
- [x] #1969
- [x] #1970

## Milestones

- [x] Copy changes: https://github.com/WordPress/openverse/milestone/18
- [x] Implementation plan drafting: https://github.com/WordPress/openverse/milestone/8
- [x] Django admin access control and basic functionality changes: https://github.com/WordPress/openverse/milestone/22
- [x] Bulk moderation actions: https://github.com/WordPress/openverse/milestone/24
- [x] ~Metrification: https://github.com/WordPress/openverse/milestone/27~

**Comment by krysal** [comment]
Required work to enable the DAG for reported media pending review created at WordPress/openverse-catalog#513. I noticed it hadn't run yet.

**Comment by sarayourfriend** [comment]
@AetherUnbound has started on the first implementation plan for this project. We probably need to get the next implementation plan also started. The project proposal shows that the API cache management and the initial Django access control implementation plans can be started at any time. If any @WordPress/openverse-maintainers do not currently have an implementation plan they are working on or can take on a new one, please volunteer for one of the ones for this project.

I will start working on https://github.com/WordPress/openverse/issues/1969 next week.

**Comment by sarayourfriend** [comment]
#2126 is pending review from the requested reviewers, @dhruvkb and @krysal. @zackkrida given the delay, do you think it's appropriate for my review of the implementation plan to serve as one of the two approvals for go ahead? @dhruvkb and @krysal, can y'all share an ETA of when you'll be able to review the implementation plan? It's been in the decision round much longer than the expected time span.

I was going to start working on the #1969 IP, but that hasn't happened due to recent incidents and high priority retrospective items taking up lots of time followed by MSR. I hope to be able to start that implementation plan next week.

The recent Redis memory consumption issue has brought to attention that the fine-grained cache IP will need to consider Redis memory usage.

**Comment by sarayourfriend** [comment]
After starting to work on the fine-grained cache implementation plan, I opened a discussion about our approach to caching generally: https://github.com/WordPress/openverse/discussions/2453

In that discussion we've decided to explore significantly reducing our cache TTL, potentially low enough such that needing fine-grained cache invalidation on the search endpoint would no longer be necessary. Therefore, I will stall work on the implementation plan to further investigate how changes to the cache will effect performance, especially Elasticsearch and Postgres resource consumption. This is possible now that the thumbnails service is live and we have confirmed that API instability was due almost entirely to long-running thumbnails requests.

Other implementation plans may move forward, such as the Django admin access control and baseline changes, which would unblock several other implementation plans.

**Comment by AetherUnbound** [comment]
This project is currently on hold, but I've created the issues associated with the first IP for this work as part of this milestone: https://github.com/WordPress/openverse/milestone/18

**Comment by sarayourfriend** [comment]
The project has been on hold since we started to consider the data dump project. That's done now and we've reprioritised it. I still need to catch up on how the remaining implementation plans will be prioritised, in particular, but we're also waiting to get clarification on _who_ will work the moderation queue specifically and to confirm that Django Admin, rather than needing to integrate with some other tool, is fine.

**Comment by sarayourfriend** [comment]
The implementation plans are still being prioritised. We've made no further progress that I am aware of.

**Comment by sarayourfriend** [comment]
Apologies for the lack of updates here. The project was essentially on hold during the API response time mitigation work. The implementation plans are still in the todo column with no further progress other than https://github.com/WordPress/openverse/issues/1967 is now assigned (:tada:).

I closed #1969 with an explanatory comment. I'll share a proposed alternative next week for us to consider. I chose to close the issue to avoid anyone picking it up and working on the implementation plan, just in case it ends up not being the direction we choose. Closing the issue should not be taken as an indication that the plan is changed _now_. I'm just trying to be cautious about accidentally planning complex work that we wouldn't end up needing. I don't expect this to delay things because the other parts of the project are at times more significant (like the underlying Django admin changes to support a moderation queue to begin with).

By the way, @stacimc, the implementation plan you've picked up is tightly related to https://github.com/WordPress/openverse/issues/1966. I'd suggest re-reading [the relevant part of the project proposal if you haven't yet](https://docs.openverse.org/projects/proposals/trust_and_safety/content_report_moderation/20230411-project_proposal_content_report_moderation.html#volunteer-moderator-django-admin-tools-and-access-control-requirement-2) just to make sure the scope of the bulk moderation plan is clear to you (please ask for clarification if not) and that the planned implementation is flexible enough to fit into whatever Django admin tools we build to execute the bulk moderation actions.

**Comment by sarayourfriend** [comment]
Access control and basic feature implementation plan is ready to go! The milestone with issues is here: https://github.com/WordPress/openverse/milestone/22

**Comment by sarayourfriend** [comment]
Staci's bulk actions implementation plan got approval, and she's merged it and created new and updated existing issues to reflect the plan in the IP.

Dhruv's IP for moderation metrics is still under review, awaiting a response from @AetherUnbound, who hasn't commented yet on it: https://github.com/WordPress/openverse/pull/3760

Madison, can you add a comment in the PR clarifying when you'll be able to review the plan?

**Comment by AetherUnbound** [comment]
> Dhruv's IP for moderation metrics is still under review, awaiting a response from @AetherUnbound, who hasn't commented yet on it: #3760
> 
> Madison, can you add a comment in the PR clarifying when you'll be able to review the plan?

I was originally randomly assigned as a reviewer when @dhruvkb initially opened the PR. I was expecting for Dhruv to provide some explanation on why reviewers were chosen (as is typically done for IPs) and had ignored the PR while he continued to work on it. While I was team lead, I didn't believe I had the capacity to be an IP reviewer, so I was not expecting to be assigned review in general. I probably should have removed myself initially when the random assignment chose me, apologies for not doing that. @dhruvkb would you still like me to be an reviewer for that IP? I can try to prioritize it over the next day or two if so. Also, do you mind updating the PR description for it to use the [proposal template](https://github.com/WordPress/openverse/blob/main/.github/PULL_REQUEST_TEMPLATE/project_proposal.md) so we can see the project-related information there?

**Comment by sarayourfriend** [comment]
Work on the moderation decision and new admin views is underway, with https://github.com/WordPress/openverse/issues/3634 recently closed to add the new admin view for user preferences.

We've also continued discussions behind the scenes regarding who will do the moderation. It is my understanding we have an answer on this question, but I'll ping @zackkrida here to clarify if that is the case. If we now have a definitive answer, do you think it's worth documenting that in this project thread?

I've been focused almost exclusively on https://github.com/WordPress/openverse/issues/2037, but folks have been picking up issues and reviewing IPs here and there. We only have one IP left to write for this, and once the openverse.org migration for the API and Airflow is finished, I'll be able to put some more time towards helping implement this project.

**Comment by zackkrida** [comment]
> We've also continued discussions behind the scenes regarding who will do the moderation. It is my understanding we have an answer on this question, but I'll ping @zackkrida here to clarify if that is the case. If we now have a definitive answer, do you think it's worth documenting that in this project thread?

We don't yet have an answer but should by the end of this month. I will update the project thread here and ping you when that occurs!

**Comment by zackkrida** [comment]
@sarayourfriend We will move forward with a sub-set of volunteers from @WordPress/openverse-maintainers handling content moderation.

**Comment by sarayourfriend** [comment]
Apologies for missing an update on this.

@dhruvkb's got a very cool PR for the "locking" of reports, so that moderators know when someone is reviewing a work or not: https://github.com/WordPress/openverse/pull/4374

I also opened a related issue, that will sit outside the project, for us to consider different ways of logging into Django admin that could make user management and permissions a bit easier: https://github.com/WordPress/openverse/issues/4342. That is out of scope for this project, and we do not realistically need, nor would benefit from all that complexity. The plan right now does not involved a large number of moderators, nor would they change often, so we don't benefit from making user management overall easier, if the underlying implementation is complicated.

I intended to start working on some issues for this project this week, but got pulled away to things related to an API incident. Hopefully I'll wrap those up early next week and be able to work on some of the issues from this milestone: https://github.com/WordPress/openverse/milestone/22

Krystle's picked up the computer vision labelling implementation plan. Looking forward to seeing that one, especially with the ongoing work in #431 in the works.

**Comment by sarayourfriend** [comment]
To reduce scope and help make more time/room in our schedule for projects important to search, we're cutting the computer vision and metrification aspects of this project. They were always considered "nice to have", and while I'm disappointed @dhruvkb's IP for the metrification part won't be implemented right away, the work is planned and available for us to work on later. It is less critical now that we know who will be doing the moderation. The visibility we'll have with the basic features after we've implemented bulk moderation will be more than sufficient to ensure reports are reviewed in a timely and consistent manner, as monitored by the moderators themselves (following self-accountability, if it needs to be more explicit).

The computer vision aspect would also be nice to have, and a continuation of #4040, but ultimately not necessary for MVP.

**Comment by zackkrida** [comment]
> @zackkrida do you think we should close the issues in the milestone, or leave them as is?

I've just peeked at the milestones for the project. I closed the IP milestone, and the two remaining milestones look to contain the correct issues.  

Django admin access control and basic functionality changes: [WordPress/openverse/milestone/22](https://github.com/WordPress/openverse/milestone/22)
Bulk moderation actions: [WordPress/openverse/milestone/24](https://github.com/WordPress/openverse/milestone/24)

Perhaps you've already made changes, @sarayourfriend? In any case, it all looks good to me.

**Comment by sarayourfriend** [comment]
I meant specifically about the metrification issues. There are only three, and they're arguably unnecessary with the current plan of who will do the moderation. I'll close them as won't do for now, and we can revisit our needs in the future once we actually need it.

**Comment by zackkrida** [comment]
Ah thanks! I saw that those weren't part of the project's milestones so figured they'd be acceptable to keep as low priority for a rainy day, but you're totally right given our set of moderators.

**Comment by sarayourfriend** [comment]
Oops! I thought I'd left an update here two weeks ago! Apologies.

A single issue remains in the initial milestone, https://github.com/WordPress/openverse/issues/3640, assigned to @dhruvkb. That one does not block work on the bulk moderation issues, so we could start on those issues as soon as we like.

The backfill had one small issue with foreign keys, fixed in https://github.com/WordPress/openverse/issues/4512. We ran it in production, and have now dropped the status column in production. The new model for report decision tracking is fully operational!

**Comment by zackkrida** [comment]
@dhruvkb do you mind contributing this update while @sarayourfriend is AFK?

**Comment by dhruvkb** [comment]
Issue #3640 from milestone https://github.com/WordPress/openverse/milestone/22 has an associated PR #4603, that has been reviewed by @sarayourfriend and the changes that Sara requested have been addressed. Since Sara is AFK, I have requested a review from `WordPress/openverse-api` team.

Also I have started working on issue #3840 from milestone https://github.com/WordPress/openverse/milestone/24 and a PR for that will be up very soon.

**Comment by dhruvkb** [comment]
All milestones for this project have now been completed, so the project has been moved to the "🚢 Shipped" column.

**Comment by zackkrida** [comment]
@WordPress/openverse-maintainers this project will be marked a success after the tooling has actually been used in a two-week trial by the sub-group of maintainers who will perform moderation. 

Feedback from this trial may inform further improvements or changes.

**Comment by sarayourfriend** [comment]
It would be great to split feedback between bugs with the functionality implemented for this project and with iterative improvements and additional features, just so we don't accidentally expand the scope of this project beyond its original goals for the foundational functionality.

**Comment by sarayourfriend** [comment]
@zackkrida do you know when the two-week trial will start, and if so, can you share the details here? If not, how can I help get that moving forward?

Nothing to add in the way of an update beyond that, though.

**Comment by zackkrida** [comment]
@sarayourfriend given upcoming team AFK and travel, I think the trial should take place from October 14-25th.

Additionally, I think we should stop sending reminder pings for shipped projects:
https://github.com/WordPress/openverse/issues/4912

**Comment by sarayourfriend** [comment]
See https://github.com/WordPress/openverse/issues/383#issuecomment-2344133502

Also, I opened a PR to stop these pings on shipped projects as suggested by Zack: #4993.
